Invalid segment path when using HDFS as the intermediate deepstore

### Affected Version

Druid 27

### Description

#### The cluster is deployed on k8s with following features enabled
1. MM less feature
2. Use HDFS as deepstore
3. Set `druid_processing_intermediaryData_storage_type: "deepstore"`

#### Steps to reproduce the problem
1. Submit an `index_parallel` type batch task
2. Set `maxNumConcurrentSubTasks` > 1

#### Error logs
```
Caused by: java.lang.IllegalArgumentException: Pathname /user/druid/k8s/druid27-test/segments/shuffle-data/index_parallel_session_analysis_test_bkdhhedd_2023-09-08T03:18:21.121Z/2023-08-29T16:00:00.000Z/2023-08-29T17:00:00.000Z/0 from hdfs://R2/user/druid/k8s/druid27-test/segments/shuffle-data/index_parallel_session_analysis_test_bkdhhedd_2023-09-08T03:18:21.121Z/2023-08-29T16:00:00.000Z/2023-08-29T17:00:00.000Z/0 is not a valid DFS filename.
	at org.apache.hadoop.hdfs.DistributedFileSystem.getPathName(DistributedFileSystem.java:266) ~[?:?]
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1558) ~[?:?]
	at org.apache.hadoop.hdfs.DistributedFileSystem$29.doCall(DistributedFileSystem.java:1555) ~[?:?]
	at org.apache.hadoop.fs.FileSystemLinkResolver.resolve(FileSystemLinkResolver.java:81) ~[?:?]
	at org.apache.hadoop.hdfs.DistributedFileSystem.mkdirsInternal(DistributedFileSystem.java:1572) ~[?:?]
	at org.apache.hadoop.hdfs.DistributedFileSystem.mkdirs(DistributedFileSystem.java:1547) ~[?:?]
	at org.apache.hadoop.fs.FileSystem.mkdirs(FileSystem.java:2404) ~[?:?]
	at org.apache.druid.storage.hdfs.HdfsDataSegmentPusher.pushToPath(HdfsDataSegmentPusher.java:154) ~[?:?]
	at org.apache.druid.indexing.worker.shuffle.DeepStorageIntermediaryDataManager.addSegment(DeepStorageIntermediaryDataManager.java:78) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.worker.shuffle.ShuffleDataSegmentPusher.push(ShuffleDataSegmentPusher.java:66) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.segment.realtime.appenderator.AppenderatorImpl.lambda$mergeAndPush$4(AppenderatorImpl.java:956) ~[druid-server-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:129) ~[druid-processing-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:81) ~[druid-processing-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:163) ~[druid-processing-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.java.util.common.RetryUtils.retry(RetryUtils.java:153) ~[druid-processing-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.segment.realtime.appenderator.AppenderatorImpl.mergeAndPush(AppenderatorImpl.java:952) ~[druid-server-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.segment.realtime.appenderator.AppenderatorImpl.lambda$push$1(AppenderatorImpl.java:784) ~[druid-server-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at com.google.common.util.concurrent.Futures$1.apply(Futures.java:713) ~[guava-16.0.1.jar:?]
	at com.google.common.util.concurrent.Futures$ChainingListenableFuture.run(Futures.java:861) ~[guava-16.0.1.jar:?]
	... 3 more
```

```
2023-09-08T07:07:32,491 ERROR [task-runner-0-priority-0] org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner - Exception while running task[AbstractTask{id='partial_index_generic_merge_session_analysis_test_lmhgjiib_2023-09-08T07:07:16.638Z', groupId='index_parallel_session_analysis_test_mohoklnj_2023-09-08T07:03:43.955Z', taskResource=TaskResource{availabilityGroup='partial_index_generic_merge_session_analysis_test_lmhgjiib_2023-09-08T07:07:16.638Z', requiredCapacity=1}, dataSource='session_analysis_test', context={forceTimeChunkLock=true, useLineageBasedSegmentAllocation=true}}]
java.io.IOException: org.apache.druid.segment.loading.SegmentLoadingException: Do not know how to handle file type at [hdfs://R2/user/druid/k8s/druid27-test/segments/shuffle-data/index_parallel_session_analysis_test_mohoklnj_2023-09-08T07_03_43.955Z/20230830T060000.000Z_20230830T070000.000Z/0/partial_index_generate_session_analysis_test_fgbidneh_2023-09-08T07_04_41.368Z]
	at org.apache.druid.indexing.common.task.batch.parallel.DeepStorageShuffleClient.fetchSegmentFile(DeepStorageShuffleClient.java:56) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.common.task.batch.parallel.DeepStorageShuffleClient.fetchSegmentFile(DeepStorageShuffleClient.java:33) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.common.task.batch.parallel.PartialSegmentMergeTask.fetchSegmentFiles(PartialSegmentMergeTask.java:228) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.common.task.batch.parallel.PartialSegmentMergeTask.runTask(PartialSegmentMergeTask.java:173) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.common.task.batch.parallel.PartialGenericSegmentMergeTask.runTask(PartialGenericSegmentMergeTask.java:46) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.common.task.AbstractTask.run(AbstractTask.java:173) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:477) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at org.apache.druid.indexing.overlord.SingleTaskBackgroundRunner$SingleTaskBackgroundRunnerCallable.call(SingleTaskBackgroundRunner.java:449) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	at java.util.concurrent.FutureTask.run(FutureTask.java:264) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) ~[?:?]
	at java.lang.Thread.run(Thread.java:829) ~[?:?]
Caused by: org.apache.druid.segment.loading.SegmentLoadingException: Do not know how to handle file type at [hdfs://R2/user/druid/k8s/druid27-test/segments/shuffle-data/index_parallel_session_analysis_test_mohoklnj_2023-09-08T07_03_43.955Z/20230830T060000.000Z_20230830T070000.000Z/0/partial_index_generate_session_analysis_test_fgbidneh_2023-09-08T07_04_41.368Z]
	at org.apache.druid.storage.hdfs.HdfsDataSegmentPuller.getSegmentFiles(HdfsDataSegmentPuller.java:288) ~[?:?]
	at org.apache.druid.storage.hdfs.HdfsLoadSpec.loadSegment(HdfsLoadSpec.java:57) ~[?:?]
	at org.apache.druid.indexing.common.task.batch.parallel.DeepStorageShuffleClient.fetchSegmentFile(DeepStorageShuffleClient.java:52) ~[druid-indexing-service-27.0.0-SNAPSHOT.jar:27.0.0-SNAPSHOT]
	... 11 more
```

### RCA

The root cause of the issue is that we call method `pushToPath` passing a suffix dir
https://github.com/apache/druid/blob/286eecad7cfe02369157073fd4915a7a0b480ec1/indexing-service/src/main/java/org/apache/druid/indexing/worker/shuffle/DeepStorageIntermediaryDataManager.java#L78
While method [HdfsDataSegmentPusher#pushToPath](https://github.com/apache/druid/blob/23b78c0f955860d51add780178ef8a4526b334db/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java#L119) treats parameter `storageDirSuffix` as output index fiilepath with "index.zip" suffix, see 
https://github.com/apache/druid/blob/23b78c0f955860d51add780178ef8a4526b334db/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java#L128
https://github.com/apache/druid/blob/23b78c0f955860d51add780178ef8a4526b334db/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java#L148-L156


### Solution

In method method [HdfsDataSegmentPusher#pushToPath], we should check if the parameter is a dir or a well-formatted segment file path.

To fix it, we can change this line
https://github.com/apache/druid/blob/23b78c0f955860d51add780178ef8a4526b334db/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java#L128
to
```java
    final String outIndexFilePath;
    if (storageDirSuffix.endsWith("index.zip")) {
      outIndexFilePath = StringUtils.format("%s/%s", fullyQualifiedStorageDirectory.get(), storageDirSuffix);
    } else {
      outIndexFilePath = StringUtils.format(
          "%s/%s/%d_index.zip",
          fullyQualifiedStorageDirectory.get(),
          storageDirSuffix.replace(':', '_'),  // We don't want to include ':' characters in hdfs path
          segment.getShardSpec().getPartitionNum()
      );
    }
```

--- a/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java
+++ b/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java
@@ -124,8 +124,18 @@ public class HdfsDataSegmentPusher implements DataSegmentPusher
         fullyQualifiedStorageDirectory.get(),
         storageDirSuffix
     );
-
-    final String storageDir = StringUtils.format("%s/%s", fullyQualifiedStorageDirectory.get(), storageDirSuffix);
+    
+    final String outIndexFilePath;
+    if (storageDirSuffix.endsWith("index.zip")) {
+      outIndexFilePath = StringUtils.format("%s/%s", fullyQualifiedStorageDirectory.get(), storageDirSuffix);
+    } else {
+      outIndexFilePath = StringUtils.format(
+          "%s/%s/%d_index.zip",
+          fullyQualifiedStorageDirectory.get(),
+          storageDirSuffix.replace(':', '_'),
+          segment.getShardSpec().getPartitionNum()
+      );
+    }
 
     Path tmpIndexFile = new Path(StringUtils.format(
         "%s/%s/%s/%s_index.zip",
@@ -145,7 +155,7 @@ public class HdfsDataSegmentPusher implements DataSegmentPusher
       try (FSDataOutputStream out = fs.create(tmpIndexFile)) {
         size = CompressionUtils.zip(inDir, out);
       }
-      final Path outIndexFile = new Path(storageDir);
+      final Path outIndexFile = new Path(outIndexFilePath);
       dataSegment = segment.withLoadSpec(makeLoadSpec(outIndexFile.toUri()))
                            .withSize(size)
                            .withBinaryVersion(SegmentUtils.getVersionFromDir(inDir));

--- a/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java
+++ b/extensions-core/hdfs-storage/src/main/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusher.java
@@ -106,36 +106,37 @@ public class HdfsDataSegmentPusher implements DataSegmentPusher
 
     final String uniquePrefix = useUniquePath ? DataSegmentPusher.generateUniquePath() + "_" : "";
     final String outIndexFilePathSuffix = StringUtils.format(
-        "%s/%d_%sindex.zip",
+        "%s/%s/%d_%sindex.zip",
+        fullyQualifiedStorageDirectory.get(),
         storageDir,
         segment.getShardSpec().getPartitionNum(),
         uniquePrefix
     );
 
-    return pushToPath(inDir, segment, outIndexFilePathSuffix);
+    return pushToFilePath(inDir, segment, outIndexFilePathSuffix);
   }
 
   @Override
   public DataSegment pushToPath(File inDir, DataSegment segment, String storageDirSuffix) throws IOException
   {
+    String outIndexFilePath = StringUtils.format(
+        "%s/%s/%d_index.zip",
+        fullyQualifiedStorageDirectory.get(),
+        storageDirSuffix.replace(':', '_'),
+        segment.getShardSpec().getPartitionNum()
+    );
+    
+    return pushToFilePath(inDir, segment, outIndexFilePath);
+  }
+  
+  private DataSegment pushToFilePath(File inDir, DataSegment segment, String outIndexFilePath) throws IOException
+  {
     log.debug(
         "Copying segment[%s] to HDFS at location[%s/%s]",
         segment.getId(),
         fullyQualifiedStorageDirectory.get(),
-        storageDirSuffix
+        outIndexFilePath
     );
-    
-    final String outIndexFilePath;
-    if (storageDirSuffix.endsWith("index.zip")) {
-      outIndexFilePath = StringUtils.format("%s/%s", fullyQualifiedStorageDirectory.get(), storageDirSuffix);
-    } else {
-      outIndexFilePath = StringUtils.format(
-          "%s/%s/%d_index.zip",
-          fullyQualifiedStorageDirectory.get(),
-          storageDirSuffix.replace(':', '_'),
-          segment.getShardSpec().getPartitionNum()
-      );
-    }
 
     Path tmpIndexFile = new Path(StringUtils.format(
         "%s/%s/%s/%s_index.zip",

--- a/extensions-core/hdfs-storage/src/test/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusherTest.java
+++ b/extensions-core/hdfs-storage/src/test/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusherTest.java
@@ -208,14 +208,6 @@ public class HdfsDataSegmentPusherTest
         segment.getLoadSpec().get("path").toString(),
         segment.getLoadSpec().get("path").toString().endsWith(storageDirSuffix.replace(':', '_') + "/0_index.zip")
     );
-    
-    final String outIndexFilePathSuffix = "foo/20150101T000000.000Z_20160101T000000.000Z/0/0_index.zip";
-    segment = pusher.pushToPath(segmentDir, segmentToPush, outIndexFilePathSuffix);
-
-    Assert.assertTrue(
-        segment.getLoadSpec().get("path").toString(),
-        segment.getLoadSpec().get("path").toString().endsWith(outIndexFilePathSuffix)
-    );
   }
 
   private void testUsingSchemeForMultipleSegments(final String scheme, final int numberOfSegments) throws Exception

--- a/extensions-core/hdfs-storage/src/test/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusherTest.java
+++ b/extensions-core/hdfs-storage/src/test/java/org/apache/druid/storage/hdfs/HdfsDataSegmentPusherTest.java
@@ -170,6 +170,54 @@ public class HdfsDataSegmentPusherTest
     );
   }
 
+  @Test
+  public void testPushToPath() throws Exception
+  {
+    Configuration conf = new Configuration(true);
+
+    // Create a mock segment on disk
+    File segmentDir = tempFolder.newFolder();
+    File tmp = new File(segmentDir, "version.bin");
+
+    final byte[] data = new byte[]{0x0, 0x0, 0x0, 0x1};
+    Files.write(data, tmp);
+    final long size = data.length;
+
+    HdfsDataSegmentPusherConfig config = new HdfsDataSegmentPusherConfig();
+    final File storageDirectory = tempFolder.newFolder();
+
+    config.setStorageDirectory(StringUtils.format("file://%s", storageDirectory.getAbsolutePath()));
+    HdfsDataSegmentPusher pusher = new HdfsDataSegmentPusher(config, conf, new DefaultObjectMapper());
+
+    DataSegment segmentToPush = new DataSegment(
+        "foo",
+        Intervals.of("2015/2016"),
+        "0",
+        new HashMap<>(),
+        new ArrayList<>(),
+        new ArrayList<>(),
+        NoneShardSpec.instance(),
+        0,
+        size
+    );
+
+    final String storageDirSuffix = "shuffle-data/index_parallel_session_analysis_test_bkdhhedd_2023-09-08T03:18:21.121Z/2023-08-29T16:00:00.000Z/2023-08-29T17:00:00.000Z";
+    DataSegment segment = pusher.pushToPath(segmentDir, segmentToPush, storageDirSuffix);
+    
+    Assert.assertTrue(
+        segment.getLoadSpec().get("path").toString(),
+        segment.getLoadSpec().get("path").toString().endsWith(storageDirSuffix.replace(':', '_') + "/0_index.zip")
+    );
+    
+    final String outIndexFilePathSuffix = "foo/20150101T000000.000Z_20160101T000000.000Z/0/0_index.zip";
+    segment = pusher.pushToPath(segmentDir, segmentToPush, outIndexFilePathSuffix);
+
+    Assert.assertTrue(
+        segment.getLoadSpec().get("path").toString(),
+        segment.getLoadSpec().get("path").toString().endsWith(outIndexFilePathSuffix)
+    );
+  }
+
   private void testUsingSchemeForMultipleSegments(final String scheme, final int numberOfSegments) throws Exception
   {
     Configuration conf = new Configuration(true);

